### PR TITLE
sarkar/Valid sequence length for sdpa

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?

See https://github.com/HabanaAI/optimum-habana-fork/pull/315 + https://github.com/HabanaAI/optimum-habana-fork/pull/322

Test in progress

base test cmd:
```
python run_generation.py --model_name_or_path /mnt/weka/data/llama_inference/Llama-2-7b-hf/ --use_hpu_graphs --use_kv_cache --bf16 --batch_size 16 --max_input_tokens 2048 --max_new_tokens 2048 --reuse_cache --trim_logits --attn_softmax_bf16 --limit_hpu_graphs --n_iterations 3 --warm 2 --use_flash_attention --flash_attention_recompute --bucket_internal --bucket_size 128 --flash_attention_causal_mask
```
`flash_attention_causal_mask` is needed for this to take effect

with this PR
766.175 tps, 47 GB, TFT 1246ms

on main:
765.519 tps, 47 gb, TFT 1336 ms

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by libinta — COMMENTED** [review]
can we add readme change also?

**Comment by emascarenhas** [comment]
Please sync your PR with main/upstream and fix any merge conflicts. Thank you.

**Comment by ssarkar2** [comment]
> can we add readme change also?

So there is no user flag for this. Also OHF doesnt have anything mentioned in readme. since this is an internal implementation detail, should we have a readme change?

**Comment by ssarkar2** [comment]
cannot get code quality to pass, even though fixes suggested by "make style" has been made

**Comment by ssarkar2** [comment]
Error when running fp8 msmt command:
```
QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_generation.py --model_name_or_path /mnt/weka/data/llama_inference/Llama-2-7b-hf/ --use_hpu_graphs --use_kv_cache --bf16 --batch_size 1 --max_input_tokens 2048 --max_new_tokens 2048 --reuse_cache --trim_logits --attn_softmax_bf16 --limit_hpu_graphs --n_iterations 1 --warm 0 --use_flash_attention --flash_attention_recompute --bucket_internal --bucket_size 128 --flash_attention_causal_mask
```

```
 File "/root/sasarkar/clean_model_garden/optimum-habana-4/optimum-habana/optimum/habana/transformers/models/llama/modeling_llama.py", line 669, in pre_attn_forward
    attn_output = self.fused_scaled_dot_product_attention(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1535, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1585, in _call_impl
    result = forward_call(*args, **kwargs)
TypeError: PatchedModuleFusedSDPA.forward_measure() takes from 4 to 9 positional arguments but 12 were given

```

File: examples/text-generation/utils.py [hunk 1]
```diff
@@ -573,6 +573,7 @@ def setup_generation_config(args, model, assistant_model, tokenizer):
     generation_config.flash_attention_causal_mask = args.flash_attention_causal_mask
     generation_config.flash_attention_fast_softmax = args.flash_attention_fast_softmax
     generation_config.trust_remote_code = args.trust_remote_code
+    setattr(generation_config, 'valid_sequence_lengths', None)
```
Comment (jiminha) [inline examples/text-generation/utils.py:576]: @hsubramony  @ssarkar2  Isn't this where the default value is getting set? why do we get no attribute error?

Comment (regisss) [reply examples/text-generation/utils.py:576]: `valid_sequence_lengths` should be declared here: https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/configuration_utils.py

File: optimum/habana/transformers/generation/utils.py [hunk 2]
```diff
@@ -1082,6 +1082,8 @@ def generate(
             True if generation_config.flash_attention_fast_softmax else False
         )
         model_kwargs["num_virtual_tokens"] = num_virtual_tokens
+        if hasattr(generation_config, "valid_sequence_lengths"):
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1085]:
```suggestion
        if generation_config.valid_sequence_lengths is not None:
```

**Review by regisss — APPROVED** [review]

**Pull request merged**